### PR TITLE
add choose next person strategy

### PR DESCRIPTION
**Pull request by JoanWu5**

### CODE REVIEW

File: players/g1_player.py [hunk 1]
```diff
@@ -29,12 +31,80 @@ def generate_choices(top_layer: np.ndarray, curr_level: np.ndarray) -> List[Choi
...
+    return res
+
+
 class Player:
     def __init__(self, flavor_preference: List[int], rng: np.random.Generator, logger: logging.Logger) -> None:
         self.flavor_preference = flavor_preference
         self.rng = rng
         self.logger = logger
         self.state = [0]
+        self.group_id = 1 - 1
```
Comment (zerolfx) [inline players/g1_player.py]: ?

File: players/g1_player.py [hunk 2]
```diff
@@ -48,10 +118,18 @@ def serve(self, top_layer: np.ndarray, curr_level: np.ndarray, player_idx: int,
...
-            return dict(action="pass", values=self.rng.choice(players))
+            # if we choose the person with the highest score
+            next_player = choose_next_player(min(turns), players, get_served(), top_layer, curr_level, self.rng,
+                                             self.group_id)
+            return dict(action="pass", values=next_player)
+
+            # if we just randomly choose one person
+            # if len(players) == 0:
+            #     players = [i for i in range(10) if i != self.group_id]
+            # return dict(action="pass", values=self.rng.choice(players))
```
Comment (zerolfx) [inline players/g1_player.py:147]: delete or fix: only pass to players that haven't got served in this round

File: players/g1_player.py [hunk 3]
```diff
@@ -61,6 +139,7 @@ def f(choice: Choice) -> float:
             res += choice.max_depth * 0.2
             res += 0.01 * len(choice.flavors)
             return res
+
         choice = max(choices, key=f)
```
Comment (zerolfx) [inline players/g1_player.py:160]: remove duplicated function f
you may need `functools.partial`

File: players/g1_player.py [hunk 4]
```diff
@@ -29,12 +31,80 @@ def generate_choices(top_layer: np.ndarray, curr_level: np.ndarray) -> List[Choi
...
+        return rng.choice(possible_players)
+
+    # if there are no candidates, which means our team is the last one to serve
+    if len(possible_players) == 0:
+        possible_players = [i for i in range(10) if i != group_id]
+
+    choices = generate_choices(top_layer, curr_level)
+    # if all ice cream has been taken
+    if len(choices) == 0:
+        return rng.choice(list(possible_players))
```
Comment (zerolfx) [inline players/g1_player.py:51]: impossible situation
delete or replace by `assert`

File: players/g1_player.py [hunk 5]
```diff
@@ -29,12 +31,80 @@ def generate_choices(top_layer: np.ndarray, curr_level: np.ndarray) -> List[Choi
...
+
+    greedy_flavor_preference = defaultdict(list)
+    for player_index in possible_players:
+        player_served_situation = sorted(served_situation[player_index].items(), key=lambda item: -item[1])
+        for flavor, _ in player_served_situation:
+            greedy_flavor_preference[player_index].append(flavor)
+
+    max_score = -math.inf
+    next_player_list = set()
```
Comment (zerolfx) [inline players/g1_player.py]: use `list` instead?

**Review by zerolfx — CHANGES_REQUESTED** [review]

File: players/g1_player.py (hunk 1 above)
Comment (JoanWu5) [reply players/g1_player.py]: it is just our group_id, I just remind myself to minus 1

File: players/g1_player.py (hunk 2 above)
Comment (JoanWu5) [reply players/g1_player.py:147]: it is for random test
if you are the last one to pass to in this round, which means, you can pass to anyone except yourself in the next round

File: players/g1_player.py (hunk 4 above)
Comment (JoanWu5) [reply players/g1_player.py:51]: actually, I test it, and it truly exists

File: players/g1_player.py (hunk 3 above)
Comment (JoanWu5) [reply players/g1_player.py:160]: I will revise score function then, to try what you have said in wechat

File: players/g1_player.py (hunk 5 above)
Comment (JoanWu5) [reply players/g1_player.py]: okay

**Review by zerolfx — APPROVED** [review]

**Pull request merged**